coalib/output/JSONEncoder.py

```diff
@@ -21,6 +21,12 @@ def default(self, obj):
                 params = self._filter_params(
                     fdata.optional_params, fdata.non_optional_params)
                 return obj.__json__(**params)
+            elif 'package_managers' in dir(obj):
+                return {member: getattr(obj, member)
+                        for member in dir(obj)
+                        if not member == 'package_managers'
```

Comment:
`not foo == bar` ?? try `!=` ?

However, this is the wrong level to be putting a special case for 'package_managers'.  That should be specified by the caller.  Can the `__json__` attribute (see above) be used to skip the package_managers and other undesirable bits?

If not, you can add explicit keywords to `create_json_encoder` for names to be skips, or some other mechanism of bypassing the items we dont want.

Another approach is modify the container of `package_managers` to to indicate that the attribute `package_managers` does not need to be serialised, because it will automatically be recreated again.  This is the `__setstate__` and `__getstate__` methods described at https://docs.python.org/2/library/pickle.html#object.__getstate__

Reply:
> If not, you can add explicit keywords to create_json_encoder for names to be skips, or some other mechanism of bypassing the items we dont want.
I looked into this and the only way seem to override the `iterencode` method which is not very clean approach. Also, I am not sure if this would be very reusable or required at all in the future. 

> Another approach is modify the container of package_managers to to indicate that the attribute package_managers does not need to be serialised, because it will automatically be recreated again. This is the __setstate__ and __getstate__
I think this would be a better solution to solve this problem. Also, then we'd have to use the pickle module rather than json here.

Reply:
One solution would be to add the `__json__` attribute to the `DistributionRequirement` class and there avoid adding the `package_managers`.
Another solution could be that we pass a list of elements that are to be avoided to `get_public_members()`
Other solutions I could find were to override `iterencode` in `JSONEncoder` which is not very clean or reusable.
> Another approach is modify the container of package_managers to to indicate that the attribute package_managers does not need to be serialised, because it will automatically be recreated again. This is the __setstate__ and __getstate__ methods described at https://docs.python.org/2/library/pickle.html#object.__getstate__

Not sure about this. Requires using pickle we get binary serialisation. So, do we output that?